### PR TITLE
ci: update pre-commit, use ruff and fix mypy

### PR DESCRIPTION
**Pull request by tlambert03**
closes #260 

closes #272

### CODE REVIEW

**Comment by tlambert03** [comment]
since pre-commit was failing on #272, this fixes it, and also does what #272 was doing

File: pyproject.toml [hunk 1]
```diff
@@ -23,3 +23,25 @@ markers = [
...
+    "TID",  # flake8-tidy-imports
+    "RUF",  # ruff-specific rules
+]
+
+[tool.ruff.per-file-ignores]
+"npe2/cli.py" = ["B008","A00"]
+
+[tool.ruff.pyupgrade]
+# Preserve types, even if a file imports `from __future__ import annotations`.
+keep-runtime-typing = true
```
Comment (Czaki) [inline pyproject.toml:47]:
```suggestion
[tool.ruff.pyupgrade]
# Preserve types, even if a file imports `from __future__ import annotations`.
keep-runtime-typing = true

[tool.ruff.isort]
known-first-party=['npe2']
```

For proper separation of `npe2` import

**Review by Czaki — APPROVED** [review]
One suggestion. 
It looks like I do not have permissions in this repository.

File: pyproject.toml (hunk 1 above)
Comment (tlambert03) [reply pyproject.toml:47]: (just gave all core devs write permission)

File: pyproject.toml [hunk 2]
```diff
@@ -23,3 +23,28 @@ markers = [
...
+src = ["npe2", "tests"]
+select = [
+    "E", "F", "W", #flake8
+    "UP", # pyupgrade
+    "I", # isort
+    "B", # flake8-bugbear
+    "C4", # flake8-comprehensions
+    "TID",  # flake8-tidy-imports
+    "RUF",  # ruff-specific rules
+]
```
Comment (Czaki) [inline pyproject.toml:40]:
```suggestion
]
fix = true
```

enable autofix by pre-commit ci

Comment (tlambert03) [reply pyproject.toml:41]: just did it in cdf23f8

**Comment by tlambert03** [comment]
i'm about to head out.  you should have merge permissions now

**Comment by tlambert03** [comment]
merging this to try to fix https://github.com/napari/napari/pull/5592#issuecomment-1447169318

**Pull request merged**